contracts/api3-server-v1/proxies/Api3CrossRateReaderProxyV1.sol

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.27;
+
+import "../../vendor/@openzeppelin/contracts@5.0.2/proxy/utils/UUPSUpgradeable.sol";
+import "../../vendor/@openzeppelin/contracts-upgradeable@5.0.2/access/OwnableUpgradeable.sol";
+import "../../vendor/@chainlink/contracts@1.2.0/src/v0.8/shared/interfaces/AggregatorV2V3Interface.sol";
+import "./interfaces/IApi3CrossRateReaderProxyV1.sol";
+import "../../interfaces/IApi3ReaderProxy.sol";
+
+/// @title UUPS-upgradeable IApi3CrossRateReaderProxyV1 and
+/// AggregatorV2V3Interface implementation that is designed to be deployed by
+/// Api3CrossRateReaderProxyV1Factory
+/// @notice The owner of this contract is allowed to upgrade it. In the case
+/// that it is deployed by Api3CrossRateReaderProxyV1Factory, the owner will be
+/// the owner of Api3CrossRateReaderProxyV1Factory at the time of deployment.
+/// @dev For a gas-cheap `read()` implementation, this upgradeable contract
+/// uses immutable variables (rather than initializable ones). To enable this,
+/// an Api3CrossRateReaderProxyV1 needs to be deployed for each unique
+/// combination of variables. The end user does not need to concern themselves
+/// with this, as Api3CrossRateReaderProxyV1Factory abstracts this detail away.
+/// Refer to https://github.com/api3dao/migrate-from-chainlink-to-api3 for more
+/// information about the Chainlink interface implementation.
+contract Api3CrossRateReaderProxyV1 is
+    UUPSUpgradeable,
+    OwnableUpgradeable,
+    AggregatorV2V3Interface,
+    IApi3CrossRateReaderProxyV1
+{
+    /// @notice First IApi3ReaderProxy contract address
+    address public immutable override proxy1;
+
+    /// @notice Second IApi3ReaderProxy contract address
+    address public immutable override proxy2;
+
+    /// @notice Calculation type to be used for the cross rate
+    CalculationType public immutable override calculationType;
+
+    /// @notice Derived dAPI name of the cross rate as a bytes32 string
+    /// @dev For example if the two proxies are "USD/USDT" and "USDT/EUR", the
+    /// derived dAPI name should be "USD/EUR". This is just for informational
+    /// purposes and is not used in any calculations
+    bytes32 public immutable override crossRateDapiName;
+
+    /// @dev Parameters are validated by Api3CrossRateReaderProxyV1Factory
+    /// @param proxy1_ First IApi3ReaderProxy contract address
+    /// @param proxy2_ Second IApi3ReaderProxy contract address
+    /// @param calculationType_ Calculation type to be used for the cross rate
+    /// @param crossRateDapiName_ dAPI name of the cross rate as a bytes32 string
+    constructor(
+        address proxy1_,
+        address proxy2_,
+        CalculationType calculationType_,
+        bytes32 crossRateDapiName_
+    ) {
+        proxy1 = proxy1_;
+        proxy2 = proxy2_;
+        calculationType = calculationType_;
+        crossRateDapiName = crossRateDapiName_;
```

Comment:
There are only two combinator types - multiple and divide. I'd prefer two separate contracts for that, maybe `ProductCompositeApi3ReaderProxyV1` and `QuotientCompositeApi3ReaderProxyV1`

Reply:
Fwiw there is also the `1/proxy` case

Reply:
It seems that I should only write a single `ProductCompositeApi3ReaderProxyV1` for now 🤔

Reply:
> Fwiw there is also the 1/proxy case

That can be implemented using a `ProductCompositeApi3ReaderProxyV1` and a contract that returns scalar value 1. This is also something we see being used and might add now.

Reply:
* Renamed to [ProductApi3ReaderProxyV1](https://github.com/api3dao/contracts/pull/472/commits/fd6b0832f0237a8eab7023dea781e7194dfeac42) to keep it short
* Added [InverseApi3ReaderProxyV1](https://github.com/api3dao/contracts/pull/472/commits/cedc8ab95d84589156f2f7a87d99508fc62ad421)